src/cargo/util/toml/mod.rs

```diff
@@ -1934,6 +1944,56 @@ impl<P: ResolveToPath> DetailedTomlDependency<P> {
 
             dep.set_public(p);
         }
+
+        if let (Some(artifact), is_lib, target) = (
+            self.artifact.as_ref(),
+            self.lib.unwrap_or(false),
+            self.target.as_deref(),
+        ) {
+            if cx.config.cli_unstable().bindeps {
+                let mut artifact = Artifact::parse(artifact, is_lib, target)?;
+                if dep.kind() != DepKind::Build
+                    && artifact.target() == Some(ArtifactTarget::BuildDependencyAssumeTarget)
+                {
+                    cx.warnings.push(
+                        format!(
+                            r#"`target = "target"` in normal- or dev-dependencies has no effect ({})"#, name_in_toml
+                        )
+                        .into(),
+                    );
+                    artifact.clear_target();
+                }
+                dep.set_artifact(artifact)
+            } else {
+                let msg = format!(
+                    "`artifact = …` ignored as `-Z bindeps` is not set ({})",
+                    name_in_toml
+                );
+                cx.warnings.push(msg);
```

Comment:
> Is there a particular reason this is a warning and not an error? I would expect that if trying to build a project that specified artifact, it will not build correctly if the feature is not enabled.

I think that's me not knowing how to handle it and opting for pushing doing the right thing onto the user. I agree that it's better to fail early instead by making this an error.

325d6be implements failing early as suggested.

> Also, is there a reason this is implemented as a -Z flag instead of a cargo-features?

This was part of the RFC requirement, but I vaguely remember @alexcrichton asking something similar. That's all I know, and I am happy to change it if that's the way to go.

Reply:
Hm, let me think about it.  I would lean towards `cargo-features`, but I'm concerned there might need to be global resolver changes (particularly with something like RFC 3176).  The resolver doesn't know about cargo-features, so that might have to be a global.